### ISSUE TITLE
ZAP is malforming URLs read from an XML document

### ISSUE THREAD

**Issue by pnwright84** [zaproxy/zaproxy#8299]
### Describe the bug

Some URLs from an XML sitemap contain parameters and values. The below image shows in the manual request/response editor the correct response from the web server. 
In the bottom part of the picture you can see the spidering results showing that the URL has been manipulated/malformed when reading the URLs in the sitemap.
![ZAP Screenshot](https://github.com/zaproxy/zaproxy/assets/29858166/9a8a85a9-0b80-4a7f-9d7c-d8eb04333db6)

### Steps to reproduce the behavior

1. spider a sitemap that contains URLs with multiple parameters and values
2. review the spider results

### Expected behavior

The URLs detected in the sitemap should be correctly extracted without manipulation.

### Software versions

ZAP
Version: 2.14.0

Installed Add-ons: [[id=accessControl, version=9.0.0],
[id=alertFilters, version=19.0.0], [id=alertReport,
version=14.0.0], [id=ascanrules, version=59.0.0],
[id=attacksurfacedetector, version=1.1.4], [id=authhelper,
version=0.11.0], [id=automation, version=0.34.0],
[id=bruteforce, version=15.0.0], [id=bugtracker,
version=4.0.0], [id=callgraph, version=5.0.0], [id=callhome,
version=0.10.0], [id=commonlib, version=1.20.0],
[id=custompayloads, version=0.13.0], [id=database,
version=0.3.0], [id=diff, version=14.0.0],
[id=directorylistv1, version=7.0.0], [id=domxss,
version=18.0.0], [id=encoder, version=1.4.0], [id=exim,
version=0.8.0], [id=formhandler, version=6.5.0], [id=fuzz,
version=13.12.0], [id=gettingStarted, version=16.0.0],
[id=graaljs, version=0.5.0], [id=graphql, version=0.22.0],
[id=help, version=17.0.0], [id=hud, version=0.18.0],
[id=invoke, version=14.0.0], [id=network, version=0.13.0],
[id=oast, version=0.17.0], [id=onlineMenu, version=12.0.0],
[id=openapi, version=38.0.0], [id=postman, version=0.2.0],
[id=pscanrules, version=53.0.0], [id=quickstart,
version=43.0.0], [id=regextester, version=2.0.0],
[id=replacer, version=16.0.0], [id=reports, version=0.26.0],
[id=requester, version=7.4.0], [id=retest, version=0.8.0],
[id=retire, version=0.29.0], [id=reveal, version=7.0.0],
[id=scripts, version=44.0.0], [id=selenium,
version=15.15.0], [id=soap, version=21.0.0], [id=spider,
version=0.8.0], [id=spiderAjax, version=23.18.0], [id=sse,
version=12.0.0], [id=tips, version=12.0.0], [id=tokengen,
version=15.0.0], [id=treetools, version=8.0.0],
[id=viewstate, version=3.0.0], [id=webdriverwindows,
version=69.0.0], [id=websocket, version=30.0.0], [id=zest,
version=43.0.0]]

Operating System: Windows 11
Architecture: amd64
Java Version: Eclipse Adoptium 17.0.6
System's Locale: en_GB
Display Locale: en_GB
Format Locale: en_GB
Default Charset: windows-1252
ZAP Home Directory: C:\Users\<REMOVED>\ZAP\
ZAP Installation Directory: C:\Program Files\ZAP\Zed Attack Proxy\.\
Look and Feel: FlatLaf Light (com.formdev.flatlaf.FlatLightLaf)

### Screenshots

![ZAP Screenshot](https://github.com/zaproxy/zaproxy/assets/29858166/9a8a85a9-0b80-4a7f-9d7c-d8eb04333db6)

### Would you like to help fix this issue?

- [ ] Yes

**Comment by kingthorin** [comment]
Could you confirm that these examples are accurate?

`<loc>https://example.com/tests?data=107249&amp;1ang=en</loc>`
`https://example.com/tests?amp;lang=gen&data=107249`

**Comment by kingthorin** [comment]
I'll have to test further. I added the agreed example to to the Sitemap parser Unit Test and it parsed correctly as `https://example.com/tests?lang=en&data=107249`

It might be some other part of the Spider that's mangling it 🤷‍♂️

**Comment by kingthorin** [comment]
> This is from the log DEBUG SpiderTextParser - Canonical URL constructed using 'https://xxxxxxx/xxxx/en/details?data=121678&amp;lang=en': https://xxxxxxx/xxxx/en/details?amp;lang=en&data=121678

That must be coming from somewhere else in your app.

I put this sitemap.xml in the root of a test apache server:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
    <url>
        <loc>https://example.org/</loc>
    </url>
    <url>
        <loc>http://subdomain.example.com/</loc>
    </url>
    <url>
        <loc>relative</loc>
    </url>
    <url>
        <loc>ftp://example.com</loc>
    </url>
    <url>
	
        <loc>http://www.example.com/%C7</loc>
    </url>
	<url>
		<loc>https://example.com/tests?data=107249&amp;lang=en</loc>
    </url>
</urlset>
```

It parsed fine:
![image](https://github.com/zaproxy/zaproxy/assets/7570458/d2d1a1c8-d5fe-4212-8051-716ef8a5edb1)
`https://example.com/tests?data=107249&lang=en`

**Comment by kingthorin** [comment]
The issue does seem to be in the `UrlCanonicalizer`

```console
java.lang.AssertionError: https://example.com/tests?amp;lang=en&data=107249
Expected: is "https://example.com/tests?data=107249&amp;lang=en"
     but: was "https://example.com/tests?amp;lang=en&data=107249"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.zaproxy.addon.spider.UrlCanonicalizerUnitTest.shouldCanonicalizeURIsWithAuthority(UrlCanonicalizerUnitTest.java:121)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1541)
```

Based on this quick test:
```diff
diff --git a/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java b/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java
index 76a03e7..0479f71 100644
--- a/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java
+++ b/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java
@@ -113,7 +113,7 @@
     @Test
     void shouldCanonicalizeURIsWithAuthority() {
         // Given
-        String[] uris = {"http://example.com/", "https://example.com/", "ftp://example.com/"};
+        String[] uris = {"http://example.com/", "https://example.com/", "ftp://example.com/", "https://example.com/tests?data=107249&amp;lang=en"};
         for (String uri : uris) {
             // When
             String canonicalizedUri = UrlCanonicalizer.getCanonicalUrl(ctx, uri, BASE_URL);
```

**Comment by kingthorin** [comment]
@thc202 / @psiinon 

Do you feel this is a workable solution or should I be escaping all HTML entities?

```diff
diff --git a/addOns/spider/src/main/java/org/zaproxy/addon/spider/UrlCanonicalizer.java b/addOns/spider/src/main/java/org/zaproxy/addon/spider/UrlCanonicalizer.java
index fa1ef23..116a00c 100644
--- a/addOns/spider/src/main/java/org/zaproxy/addon/spider/UrlCanonicalizer.java
+++ b/addOns/spider/src/main/java/org/zaproxy/addon/spider/UrlCanonicalizer.java
@@ -419,11 +419,11 @@
      * @return a sorted set with all parameters, or {@code null} if the query string is {@code null}
      *     or empty.
      */
-    private static SortedSet<QueryParameter> createSortedParameters(final String queryString) {
+    private static SortedSet<QueryParameter> createSortedParameters(String queryString) {
         if (queryString == null || queryString.isEmpty()) {
             return null;
         }
-
+        queryString = queryString.replaceAll("(?i)&amp;", "&");
         final String[] pairs = queryString.split("&");
         final SortedSet<QueryParameter> params = new TreeSet<>();
 
diff --git a/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java b/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java
index 76a03e7..41e59c9 100644
--- a/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java
+++ b/addOns/spider/src/test/java/org/zaproxy/addon/spider/UrlCanonicalizerUnitTest.java
@@ -34,6 +34,7 @@
 import org.junit.jupiter.api.BeforeEach;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.CsvSource;
 import org.junit.jupiter.params.provider.ValueSource;
 import org.zaproxy.addon.spider.SpiderParam.HandleParametersOption;
 import org.zaproxy.addon.spider.parser.ParseContext;
@@ -110,16 +111,16 @@
         assertThat(canonicalizedUri, is(equalTo("http://example.com/")));
     }
 
-    @Test
-    void shouldCanonicalizeURIsWithAuthority() {
-        // Given
-        String[] uris = {"http://example.com/", "https://example.com/", "ftp://example.com/"};
-        for (String uri : uris) {
-            // When
-            String canonicalizedUri = UrlCanonicalizer.getCanonicalUrl(ctx, uri, BASE_URL);
-            // Then
-            assertThat(canonicalizedUri, canonicalizedUri, is(equalTo(uri)));
-        }
+    @ParameterizedTest
+    @CsvSource({"http://example.com/, http://example.com/", 
+        "https://example.com/, https://example.com/", 
+        "ftp://example.com/, ftp://example.com/", 
+        "https://example.com/tests?data=107249&amp;lang=en, https://example.com/tests?data=107249&lang=en"})
+    void shouldCanonicalizeURIsWithAuthority(String uri, String expected) {
+        // Given / When
+        String canonicalizedUri = UrlCanonicalizer.getCanonicalUrl(ctx, uri, BASE_URL);
+        // Then
+        assertThat(canonicalizedUri, is(equalTo(expected)));
     }
 
     @Test
```

**Comment by thc202** [comment]
The `UrlCanonicalizer` is behaving as expected, it's handling `amp;` as a parameter name (which it is). (The `UrlCanonicalizer` should not be seeing XML entities in the first place.)

This is the most likely cause https://github.com/zaproxy/zaproxy/issues/8299#issuecomment-1888116773 would be good to know the actual source of those URLs.

**Comment by kingthorin** [comment]
Okay, that's fair. 👍

**Comment by kingthorin** [comment]
Thanks, I'll give it a shot.

**Comment by kingthorin** [comment]
Okay I found the issue. The SpiderSitemapXmlParser wasn't identifying this as a sitemap because of the standalone attribute in the xml tag. So it was falling through to other parsers.

I've just noticed that it also doesn't acknowledge the [OWASP sitemap](https://owasp.org/sitemap.xml) as it doesn't have an xml tag at all. Which according to <https://www.sitemaps.org/protocol.html> is actually fine it seems.

I propose we drop the XML_PATTERN regex check and depend on xpathLocationExpression.evaluate returning a `NodeList` with non-zero length.

The parser will also need to be enhanced to handle `xhtml:link href=` extraction. Although per sitemaps.org those aren't part of the protocol/spec. I'll have to read up further on those. https://developers.google.com/search/docs/crawling-indexing/sitemaps/combine-sitemap-extensions does cover it.
